### PR TITLE
Issue-200: Refactor holiday search common logic called in use cases

### PR DESCRIPTION
**Pull request by ptrimble**
The purpose of this refactor is to split out and fully test the 'holiday search' logic that makes up the core of the gem. It currently resides in the 'context common' code and is used in multiple use cases. I felt that it was time to split it out and pepper it with tests to make things more clear.

I have changed the 'year holiday', 'next holiday', and 'between' use cases to use the new logic. This is not entirely done...I need to finish cleaning things up and ensuring that I did not introduce any regressions.

Feedback welcome while I continue to work on this!

NOTE: this will not have any effect on outward gem behavior. This is an 'under the hood' update.

### CODE REVIEW

File: benchmark.rb [hunk 1]
```diff
@@ -1,8 +0,0 @@
-require 'benchmark'
-
-puts Benchmark.measure {
```
Comment (ptrimble) [inline benchmark.rb]: This was never intended to be left in. 😩

File: lib/holidays.rb [hunk 2]
```diff
@@ -193,15 +193,14 @@ def year_holidays(options, from_date = Date.today)
...
-      to_date = Date.new(from_date.year, 12, 31)
       regions, observed, informal = OptionFactory.parse_options.call(options)
 
       # This could be smarter but I don't have any evidence that just checking for
       # the next 12 months will cause us issues. If it does we can implement something
       # smarter here to check in smaller increments.
       date_driver_hash = UseCaseFactory.dates_driver_builder.call(from_date, from_date >> 12)
 
-      UseCaseFactory.year_holiday.call(from_date, to_date, date_driver_hash, regions, observed, informal)
+      UseCaseFactory.year_holiday.call(from_date, date_driver_hash, regions, observed, informal)
```
Comment (ptrimble) [inline lib/holidays.rb]: I decided that the `to_date` could easily be calculated in the use case itself now that most of the logic was in the new 'search' logic and not in 'common context'.

File: lib/holidays/definition/context/function_processor.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+              result = Date.civil(year, month, result)
+            rescue ArgumentError => e
+              raise Holidays::InvalidFunctionResponse.new("invalid day response from custom method call resulting in invalid date. Result: '#{result}'")
+            end
+          elsif result.nil?
+            # Do nothing. This is because some functions can return 'nil' today.
+            # I want to change this and so rather than come up with a clean
+            # implementation I'll do this so we don't throw an error in this specific
+            # situation. This should be removed once we have changed the existing
+            # custom definition functions. See https://github.com/holidays/holidays/issues/204
```
Comment (ttwo32) [inline lib/holidays/definition/context/function_processor.rb]: @ptrimble 👍 I will help you.

File: lib/holidays/finder/rules/year_range.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  module Finder
+    module Rules
+      class YearRange
+        class << self
+          BEFORE = :before
+          AFTER = :after
+          LIMITED = :limited
+          BETWEEN = :between
+
+          #TODO Can we just accept symbols here? Why accept strings?
```
Comment (ttwo32) [inline lib/holidays/finder/rules/year_range.rb]: If we use load_custom method, operator return string.So i added string operator.
If AFTER.to_s is not written, following case may occur.
`irb(main):001:0> Holidays.load_custom('definitions/jp.yaml')
irb(main):002:0> Holidays.on(Date.civil(2016,8,11), :jp)
=> []`

I have found above case when i was testing test_custom_year_range_holidays.rb.

Comment (ptrimble) [inline lib/holidays/finder/rules/year_range.rb]: Right! I totally see what you are saying. I ran into a similar error, which is related to my comment above. The ranges were coming in as strings. Reading from YAML does not automatically convert it to whatever the corresponding Ruby thing is.

I'm fine leaving this as-is and removing the comment. I'll try to do it this week. Thanks for explaining.

Comment (ttwo32) [inline lib/holidays/finder/rules/year_range.rb]: @ptrimble I really appreciate your work. 👍

**Pull request merged**